priam/src/main/java/com/netflix/priam/identity/token/TokenRetriever.java

```diff
@@ -155,30 +155,53 @@ public PriamInstance retriableCall() throws Exception {
-                int regionOffset = tokenManager.regionOffset(myRegion);
-                String myRac = myInstanceInfo.getRac();
-                List<String> racs = config.getRacs();
-                int mySlot =
-                        factory.getAllIds(config.getAppName())
-                                .stream()
-                                .filter(i -> i.getRac().equals(myRac))
-                                .map(PriamInstance::getId)
-                                .max(Integer::compareTo)
-                                .map(id -> racs.size() + Math.max(id, regionOffset) - regionOffset)
-                                .orElseGet(
-                                        () -> {
-                                            Preconditions.checkState(racs.contains(myRac));
-                                            return racs.indexOf(myRac);
-                                        });
-                int instanceCount = membership.getRacCount() * membership.getRacMembershipSize();
-                String newToken = tokenManager.createToken(mySlot, instanceCount, myRegion);
-                return createToken(mySlot + regionOffset, newToken);
+                return generateNewToken();
             }
         }.call();
     }
 
+    @VisibleForTesting
+    PriamInstance generateNewToken() {
+        String myRegion = myInstanceInfo.getRegion();
+        // this offset ensures the nodes are spread far away from the other regions.
+        int regionOffset = tokenManager.regionOffset(myRegion);
+        String myRac = myInstanceInfo.getRac();
+        List<String> racs = config.getRacs();
+        ImmutableSet<PriamInstance> allIds = factory.getAllIds(config.getAppName());
+        int mySlot =
+                allIds.stream()
+                        .filter(i -> i.getRac().equals(myRac))
+                        .map(PriamInstance::getId)
+                        .max(Integer::compareTo)
+                        .map(id -> racs.size() + Math.max(id, regionOffset) - regionOffset)
+                        .orElseGet(
+                                () -> {
+                                    Preconditions.checkState(racs.contains(myRac));
+                                    return racs.indexOf(myRac);
+                                });
+        int instanceCount = membership.getRacCount() * membership.getRacMembershipSize();
+        String newToken = tokenManager.createToken(mySlot, instanceCount, myRegion);
+        while (newTokenIsADuplicate(newToken, allIds)) {
+            newToken = new BigInteger(newToken).add(BigInteger.ONE).toString();
+        }
+        return createToken(mySlot + regionOffset, newToken);
+    }
+
+    private boolean newTokenIsADuplicate(String newToken, ImmutableSet<PriamInstance> instances) {
+        for (PriamInstance priamInstance : instances) {
+            if (newToken.equals(priamInstance.getToken())) {
+                if (myInstanceInfo.getRegion().equals(priamInstance.getDC())) {
```

Comment:
I'm a little unclear as to why we throw an `IllegalStateException` in this case vs the case where the DCs are different - IIUC, duplicates across DCs never happen because of the region hash offset? So it feels that _that's_ the case we should throw vs return false

Also, I'm having trouble tracing the entire code path given the multiple impls, but is it possible that `InstanceInfo#getRegion()` will return `us-east-1` for `us-east-1` while `PriamInstance#getDC` will return `us-east`?

Reply:
> IIUC, duplicates across DCs never happen because of the region hash offset?

Duplicates across DCs can happen if the region hash is off by one and the tokens between regions are off by one in the opposite direction. That scenario can result if the cluster has been doubled using Priam's doubling facilities. The purpose of this PR is to adapt to this scenario. That said, there should never be in-region duplicates and if there are we need to fail immediately. In that case we should not increment until there are no duplicates because that will lead to a situation where tokens are one apart within the same region. That would cause a data imbalance.

> is it possible that InstanceInfo#getRegion() will return us-east-1 for us-east-1 while PriamInstance#getDC will return us-east?

This is an interesting concern, but I don't believe it is a threat. The only place we ever convert away from AWS region names is in an internal tuning use case. I can go into greater depth about that this morning if you're interested.